twine/utils.py

```diff
@@ -99,23 +102,44 @@ def get_config(path: str = "~/.pypirc") -> Dict[str, RepositoryConfig]:
     return dict(config)
 
 
+def validate_url(repository_url: str) -> None:
+    """Validate the given url for allowed schemes and components"""
+
+    # Allowed schemes are http and https, based on whether the repository
+    # supports TLS or not, and scheme and host must be present in the URL
+    validator = (
+        validators.Validator()
+        .allow_schemes("http", "https")
+        .require_presence_of("scheme", "host")
+    )
+
+    url = uri_reference(repository_url)
+    validator.validate(url)
+
+
 def get_repository_from_config(
     config_file: str, repository: str, repository_url: Optional[str] = None
 ) -> RepositoryConfig:
     # Get our config from, if provided, command-line values for the
     # repository name and URL, or the .pypirc file
-    if repository_url and "://" in repository_url:
+
+    if repository_url:
+        try:
+            validate_url(repository_url)
+        except rfc_exceptions.MissingComponentError as exc:
+            raise exceptions.UnreachableRepositoryURLDetected(
+                "Repository URL has missing components. {}.".format(exc.args[0])
+            )
+        except rfc_exceptions.UnpermittedComponentError as exc:
+            raise exceptions.UnreachableRepositoryURLDetected(
+                "Repository URL has an invalid scheme. {}".format(exc.args[0])
+            )
```

Comment:
Given that the `rfc3986` exceptions already include a helpful message, and they derive from a common base, I think this could be simplified to something like:

```suggestion
        except rfc3986.exceptions.RFC3986Exception as exc:
            raise exceptions.UnreachableRepositoryURLDetected(
                f"Invalid repository URL: {exc}"
            )
```

Reply:
I have added the exception string `f"Invalid repository URL: {exc.args[0]}"` in the latest commit, since the first value of `exc.args` contain the error message